### ISSUE TITLE
NDO simple playbook with httpapi failing with Remote User

### ISSUE THREAD

**Issue by soumukhe** [CiscoDevNet/ansible-mso#220]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Versions:
ND Version: tried both vND 2.1.1e and vND 2.1.1d
NDO Version: tried NDO 3.5.1e and also NDO 3.6.0.88

(vEnvAnsible2.9.6-P3.7.9) aciadmin@ubuntu-jump:~/Ansible/ansible-scripts/MSO_NI6.0Test/testing$ ansible --version
```
ansible 2.9.6
config file = /etc/ansible/ansible.cfg
configured module search path = ['/home/aciadmin/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
ansible python module location = /home/aciadmin/Ansible/envs/vEnvAnsible2.9.6-P3.7.9/lib/python3.7/site-packages/ansible
executable location = /home/aciadmin/Ansible/envs/vEnvAnsible2.9.6-P3.7.9/bin/ansible
python version = 3.7.9 (default, Aug 24 2020, 23:58:02) [GCC 7.5.0]
```
### Description:
NDO simple playbook with httpapi failing with Remote User

### Details:
**inventory file:**
(vEnvAnsible2.9.6-P3.7.9) aciadmin@ubuntu-jump:~/Ansible/ansible-scripts/MSO_NI6.0Test/testing$ **cat hosts**
```
[nd]
nd1 ansible_host=10.0.0.94

[nd:vars]
ansible_connection=ansible.netcommon.httpapi
ansible_network_os=cisco.nd.nd
#ansible_network_os=cisco.mso.mso
ansible_httpapi_validate_certs=False
ansible_httpapi_use_ssl=True
ansible_httpapi_use_proxy=True
ansible_python_interpreter=/usr/bin/python3
```
**Simple Playbook _(with remote user)_:**

(vEnvAnsible2.9.6-P3.7.9) aciadmin@ubuntu-jump:~/Ansible/ansible-scripts/MSO_NI6.0Test/testing$ **cat simpleQueryNDO_remoteUser.yaml** 
```
- hosts: nd
  gather_facts: no
  connection: httpapi

  tasks:
  - name: Get MSO version from MSO >= 3.2
    cisco.mso.mso_version:
      host:  "nd"
      username: "soumukhe"
      password: "soumu101"
      validate_certs: "no"
      #use_proxy: false
      login_domain: "raddb"
      #port: 443
      state: query
    #delegate_to: localhost
    #register: query_result

# usage:  ansible-playbook simpleQueryNDO_remoteUser.yaml -i hosts -vvv
```

**Results (failed):  _(Note, that I am trying with remote credentials of soumukhe/soumu101 using radius domain which is verified to be correct)_**

(vEnvAnsible2.9.6-P3.7.9) aciadmin@ubuntu-jump:~/Ansible/ansible-scripts/MSO_NI6.0Test/testing$ **ansible-playbook simpleQueryNDO_remoteUser.yaml -i hosts -vvv**
```
ansible-playbook 2.9.6
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/aciadmin/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/aciadmin/Ansible/envs/vEnvAnsible2.9.6-P3.7.9/lib/python3.7/site-packages/ansible
  executable location = /home/aciadmin/Ansible/envs/vEnvAnsible2.9.6-P3.7.9/bin/ansible-playbook
  python version = 3.7.9 (default, Aug 24 2020, 23:58:02) [GCC 7.5.0]
Using /etc/ansible/ansible.cfg as config file
host_list declined parsing /home/aciadmin/Ansible/ansible-scripts/MSO_NI6.0Test/testing/hosts as it did not pass its verify_file() method
script declined parsing /home/aciadmin/Ansible/ansible-scripts/MSO_NI6.0Test/testing/hosts as it did not pass its verify_file() method
auto declined parsing /home/aciadmin/Ansible/ansible-scripts/MSO_NI6.0Test/testing/hosts as it did not pass its verify_file() method
Parsed /home/aciadmin/Ansible/ansible-scripts/MSO_NI6.0Test/testing/hosts inventory source with ini plugin

PLAYBOOK: simpleQueryNDO_remoteUser.yaml ******************************************************************************************************************************************************************************************************
1 plays in simpleQueryNDO_remoteUser.yaml

PLAY [nd] *************************************************************************************************************************************************************************************************************************************
META: ran handlers

TASK [Get MSO version from MSO >= 3.2] ********************************************************************************************************************************************************************************************************
task path: /home/aciadmin/Ansible/ansible-scripts/MSO_NI6.0Test/testing/simpleQueryNDO_remoteUser.yaml:6
<10.0.0.94> ESTABLISH LOCAL CONNECTION FOR USER: aciadmin
<10.0.0.94> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/aciadmin/.ansible/tmp/ansible-local-24833hv3y25ph/ansible-tmp-1638214621.7465324-271172996181074 `" && echo ansible-tmp-1638214621.7465324-271172996181074="` echo /home/aciadmin/.ansible/tmp/ansible-local-24833hv3y25ph/ansible-tmp-1638214621.7465324-271172996181074 `" ) && sleep 0'
Using module file /home/aciadmin/.ansible/collections/ansible_collections/cisco/mso/plugins/modules/mso_version.py
<10.0.0.94> PUT /home/aciadmin/.ansible/tmp/ansible-local-24833hv3y25ph/tmpl6nz28gn TO /home/aciadmin/.ansible/tmp/ansible-local-24833hv3y25ph/ansible-tmp-1638214621.7465324-271172996181074/AnsiballZ_mso_version.py
<10.0.0.94> EXEC /bin/sh -c 'chmod u+x /home/aciadmin/.ansible/tmp/ansible-local-24833hv3y25ph/ansible-tmp-1638214621.7465324-271172996181074/ /home/aciadmin/.ansible/tmp/ansible-local-24833hv3y25ph/ansible-tmp-1638214621.7465324-271172996181074/AnsiballZ_mso_version.py && sleep 0'
<10.0.0.94> EXEC /bin/sh -c '/usr/bin/python3 /home/aciadmin/.ansible/tmp/ansible-local-24833hv3y25ph/ansible-tmp-1638214621.7465324-271172996181074/AnsiballZ_mso_version.py && sleep 0'
<10.0.0.94> EXEC /bin/sh -c 'rm -f -r /home/aciadmin/.ansible/tmp/ansible-local-24833hv3y25ph/ansible-tmp-1638214621.7465324-271172996181074/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
  File "/tmp/ansible_cisco.mso.mso_version_payload_vc5mkz6t/ansible_cisco.mso.mso_version_payload.zip/ansible_collections/cisco/mso/plugins/module_utils/mso.py", line 505, in request
    info = self.connection.send_request(method, uri, json.dumps(data))
  File "/tmp/ansible_cisco.mso.mso_version_payload_vc5mkz6t/ansible_cisco.mso.mso_version_payload.zip/ansible/module_utils/connection.py", line 185, in __rpc__
    raise ConnectionError(to_text(msg, errors='surrogate_then_replace'), code=code)
fatal: [nd1]: FAILED! => {
    "changed": false,
    "current": {},
    "invocation": {
        "module_args": {
            "host": "nd",
            "login_domain": "raddb",
            "output_level": "normal",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": null,
            "state": "query",
            "timeout": 30,
            "use_proxy": true,
            "use_ssl": true,
            "username": "soumukhe",
            "validate_certs": false
        }
    },
    "msg": {
        "error": "Invalid Username/Password",
        "statusCode": 401
    }
}

PLAY RECAP ************************************************************************************************************************************************************************************************************************************
nd1                        : ok=0    changed=0    unreachable=0    **failed=1**    skipped=0    rescued=0    ignored=0 
```

# As a side note, this works fine with local user

**inventory file:**
(vEnvAnsible2.9.6-P3.7.9) aciadmin@ubuntu-jump:~/Ansible/ansible-scripts/MSO_NI6.0Test/testing$ cat hosts

```
[nd]
nd1 ansible_host=10.0.0.94

[nd:vars]
ansible_connection=ansible.netcommon.httpapi
ansible_network_os=cisco.nd.nd
#ansible_network_os=cisco.mso.mso
ansible_httpapi_validate_certs=False
ansible_httpapi_use_ssl=True
ansible_httpapi_use_proxy=True
ansible_python_interpreter=/usr/bin/python3

```
**simple playbook _(with local user)_**
(vEnvAnsible2.9.6-P3.7.9) aciadmin@ubuntu-jump:~/Ansible/ansible-scripts/MSO_NI6.0Test/testing$ **cat simpleQueryNDO_localUser.yaml**
```
- hosts: nd
  gather_facts: no
  connection: httpapi

  tasks:
  - name: Get MSO version from MSO >= 3.2
    cisco.mso.mso_version:
      host:  "nd"
      username: "ansible"
      password: "Ansible101!"
      validate_certs: "no"
      #use_proxy: false
      #login_domain: "raddb"
      #port: 443
      state: query
    #delegate_to: localhost
    #register: query_result

# usage:  ansible-playbook  simpleQueryNDO_localUser.yaml  -i hosts -vvv
```
**Results:  _(success with local user)_**

```
ansible-playbook 2.9.6
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/aciadmin/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/aciadmin/Ansible/envs/vEnvAnsible2.9.6-P3.7.9/lib/python3.7/site-packages/ansible
  executable location = /home/aciadmin/Ansible/envs/vEnvAnsible2.9.6-P3.7.9/bin/ansible-playbook
  python version = 3.7.9 (default, Aug 24 2020, 23:58:02) [GCC 7.5.0]
Using /etc/ansible/ansible.cfg as config file
host_list declined parsing /home/aciadmin/Ansible/ansible-scripts/MSO_NI6.0Test/testing/hosts as it did not pass its verify_file() method
script declined parsing /home/aciadmin/Ansible/ansible-scripts/MSO_NI6.0Test/testing/hosts as it did not pass its verify_file() method
auto declined parsing /home/aciadmin/Ansible/ansible-scripts/MSO_NI6.0Test/testing/hosts as it did not pass its verify_file() method
Parsed /home/aciadmin/Ansible/ansible-scripts/MSO_NI6.0Test/testing/hosts inventory source with ini plugin

PLAYBOOK: simpleQueryNDO_localUser.yaml *******************************************************************************************************************************************************************************************************
1 plays in simpleQueryNDO_localUser.yaml

PLAY [nd] *************************************************************************************************************************************************************************************************************************************
META: ran handlers

TASK [Get MSO version from MSO >= 3.2] ********************************************************************************************************************************************************************************************************
task path: /home/aciadmin/Ansible/ansible-scripts/MSO_NI6.0Test/testing/simpleQueryNDO_localUser.yaml:6
<10.0.0.94> ESTABLISH LOCAL CONNECTION FOR USER: aciadmin
<10.0.0.94> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/aciadmin/.ansible/tmp/ansible-local-25143xa43dfly/ansible-tmp-1638214937.425081-90799567782664 `" && echo ansible-tmp-1638214937.425081-90799567782664="` echo /home/aciadmin/.ansible/tmp/ansible-local-25143xa43dfly/ansible-tmp-1638214937.425081-90799567782664 `" ) && sleep 0'
Using module file /home/aciadmin/.ansible/collections/ansible_collections/cisco/mso/plugins/modules/mso_version.py
<10.0.0.94> PUT /home/aciadmin/.ansible/tmp/ansible-local-25143xa43dfly/tmp1m3f7kck TO /home/aciadmin/.ansible/tmp/ansible-local-25143xa43dfly/ansible-tmp-1638214937.425081-90799567782664/AnsiballZ_mso_version.py
<10.0.0.94> EXEC /bin/sh -c 'chmod u+x /home/aciadmin/.ansible/tmp/ansible-local-25143xa43dfly/ansible-tmp-1638214937.425081-90799567782664/ /home/aciadmin/.ansible/tmp/ansible-local-25143xa43dfly/ansible-tmp-1638214937.425081-90799567782664/AnsiballZ_mso_version.py && sleep 0'
<10.0.0.94> EXEC /bin/sh -c '/usr/bin/python3 /home/aciadmin/.ansible/tmp/ansible-local-25143xa43dfly/ansible-tmp-1638214937.425081-90799567782664/AnsiballZ_mso_version.py && sleep 0'
<10.0.0.94> EXEC /bin/sh -c 'rm -f -r /home/aciadmin/.ansible/tmp/ansible-local-25143xa43dfly/ansible-tmp-1638214937.425081-90799567782664/ > /dev/null 2>&1 && sleep 0'
ok: [nd1] => {
    "changed": false,
    "current": {
        "id": "614fcbbb1a0000a400f5c3fa",
        "platformVersion": "2.1.1d",
        "schemaversion": "",
        "timestamp": "2021-09-26T01:24:11.912149121",
        "version": "3.5.1a"
    },
    "invocation": {
        "module_args": {
            "host": "nd",
            "login_domain": null,
            "output_level": "normal",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": null,
            "state": "query",
            "timeout": 30,
            "use_proxy": true,
            "use_ssl": true,
            "username": "ansible",
            "validate_certs": false
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ************************************************************************************************************************************************************************************************************************************
nd1                        : ok=1    changed=0    unreachable=0    **failed=0**    skipped=0    rescued=0    ignored=0  
```

**Comment by soumukhe** [comment]
Lionel,  Thank You for fixing this !!!   I tested this out and it works fine now.

# Note from Lionel:
**Fixed in ND collection release v0.1.1**
 
You can update by running:
**ansible-galaxy collection install cisco.nd --force**

**Comment by lhercot** [comment]
Thanks for letting us know that this issue is resolved.
